App crashes when client disconnects abruptly

It happened whilst syncing with people and I terminated the client. Logs on a Nexus 4:

```
05-29 12:05:40.309 24256-28585/tech.relaycorp.courier I/UndispatchedCoroutine: Server started
05-29 12:05:57.591 24256-24271/tech.relaycorp.courier I/art: Background partial concurrent mark sweep GC freed 86735(4MB) AllocSpace objects, 80(28MB) LOS objects, 36% free, 27MB/43MB, paused 1.251ms total 167.709ms
05-29 12:06:13.438 24256-28698/tech.relaycorp.courier I/connections: Transport failed
    java.net.SocketException: recvfrom failed: ECONNRESET (Connection reset by peer)
        at libcore.io.IoBridge.maybeThrowAfterRecvfrom(IoBridge.java:592)
        at libcore.io.IoBridge.recvfrom(IoBridge.java:568)
        at java.nio.SocketChannelImpl.readImpl(SocketChannelImpl.java:342)
        at java.nio.SocketChannelImpl.read(SocketChannelImpl.java:304)
        at io.netty.buffer.PooledByteBuf.setBytes(PooledByteBuf.java:253)
        at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1133)
        at io.netty.channel.socket.nio.NioSocketChannel.doReadBytes(NioSocketChannel.java:350)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:148)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:615)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:578)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:818)
     Caused by: android.system.ErrnoException: recvfrom failed: ECONNRESET (Connection reset by peer)
        at libcore.io.Posix.recvfromBytes(Native Method)
        at libcore.io.Posix.recvfrom(Posix.java:177)
        at libcore.io.BlockGuardOs.recvfrom(BlockGuardOs.java:245)
        at libcore.io.IoBridge.recvfrom(IoBridge.java:565)
        at java.nio.SocketChannelImpl.readImpl(SocketChannelImpl.java:342) 
        at java.nio.SocketChannelImpl.read(SocketChannelImpl.java:304) 
        at io.netty.buffer.PooledByteBuf.setBytes(PooledByteBuf.java:253) 
        at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1133) 
        at io.netty.channel.socket.nio.NioSocketChannel.doReadBytes(NioSocketChannel.java:350) 
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:148) 
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714) 
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:615) 
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:578) 
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493) 
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) 
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) 
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) 
        at java.lang.Thread.run(Thread.java:818) 
05-29 12:06:13.489 24256-28848/tech.relaycorp.courier W/nService$deliverCargo$1: deliverCargo error
    io.grpc.StatusRuntimeException: CANCELLED: cancelled before receiving half close
        at io.grpc.Status.asRuntimeException(Status.java:524)
        at io.grpc.stub.ServerCalls$StreamingServerCallHandler$StreamingServerCallListener.onCancel(ServerCalls.java:275)
        at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.closedInternal(ServerCallImpl.java:353)
        at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.closed(ServerCallImpl.java:341)
        at io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1Closed.runInContext(ServerImpl.java:868)
        at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
        at java.lang.Thread.run(Thread.java:818)
05-29 12:06:15.719 24256-24305/tech.relaycorp.courier D/OpenGLRenderer: endAllStagingAnimators on 0xb9455708 (RippleDrawable) with handle 0xb8e12f48
05-29 12:06:15.991 24256-24256/tech.relaycorp.courier I/CogRPCServer: Server stopped
05-29 12:06:16.089 24256-24256/tech.relaycorp.courier I/art: Explicit concurrent mark sweep GC freed 46892(2MB) AllocSpace objects, 23(2MB) LOS objects, 36% free, 28MB/44MB, paused 1.403ms total 97.482ms
05-29 12:06:16.163 24256-24256/tech.relaycorp.courier I/art: Explicit concurrent mark sweep GC freed 702(25KB) AllocSpace objects, 0(0B) LOS objects, 36% free, 28MB/44MB, paused 1.159ms total 65.527ms
05-29 12:06:16.226 24256-24256/tech.relaycorp.courier I/art: Explicit concurrent mark sweep GC freed 3(96B) AllocSpace objects, 0(0B) LOS objects, 36% free, 28MB/44MB, paused 1.129ms total 62.780ms
05-29 12:06:16.278 24256-24256/tech.relaycorp.courier E/StrictMode: class tech.relaycorp.courier.ui.sync.people.PeopleSyncActivity; instances=2; limit=1
    android.os.StrictMode$InstanceCountViolation: class tech.relaycorp.courier.ui.sync.people.PeopleSyncActivity; instances=2; limit=1
        at android.os.StrictMode.setClassInstanceLimit(StrictMode.java:1)
05-29 12:06:16.279 24256-24256/tech.relaycorp.courier W/System.err: StrictMode VmPolicy violation with POLICY_DEATH; shutting down.
05-29 12:06:16.279 24256-24256/tech.relaycorp.courier I/Process: Sending signal. PID: 24256 SIG: 9
```

@sdsantos, assigning to you because this looks like a strict mode issue.

--- a/app/src/main/java/tech/relaycorp/cogrpc/server/CogRPCConnectionService.kt
+++ b/app/src/main/java/tech/relaycorp/cogrpc/server/CogRPCConnectionService.kt
@@ -21,24 +21,30 @@ class CogRPCConnectionService(
 
     override fun deliverCargo(responseObserver: StreamObserver<CargoDeliveryAck>) =
         object : StreamObserver<CargoDelivery> {
+            private var isResponseFinished = false
+
             override fun onNext(cargoDelivery: CargoDelivery) {
                 coroutineScope.launch {
                     logger.info("deliverCargo next ${cargoDelivery.id}")
                     when (serverService.deliverCargo(cargoDelivery.cargo.newInput())) {
                         CogRPCServer.DeliverResult.Successful -> {
-                            logger.info("deliverCargo next ack ${cargoDelivery.id}")
-                            responseObserver.onNext(cargoDelivery.toAck())
+                            if (!isResponseFinished) {
+                                logger.info("deliverCargo next ack ${cargoDelivery.id}")
+                                responseObserver.onNext(cargoDelivery.toAck())
+                            }
                         }
                         CogRPCServer.DeliverResult.UnavailableStorage -> {
                             logger.info("deliverCargo no space available for ${cargoDelivery.id}")
                             logger.info("deliverCargo closing with error")
                             responseObserver.onError(StatusRuntimeException(Status.RESOURCE_EXHAUSTED))
+                            isResponseFinished = true
                         }
                         CogRPCServer.DeliverResult.Invalid,
                         CogRPCServer.DeliverResult.Malformed -> {
                             logger.info("deliverCargo invalid or malformed cargo ${cargoDelivery.id}")
                             logger.info("deliverCargo closing with error")
                             responseObserver.onError(StatusRuntimeException(Status.INVALID_ARGUMENT))
+                            isResponseFinished = true
                         }
                     }
                 }

